R/find_nhd.R

```diff
@@ -0,0 +1,337 @@
+
+find_outlets <- function(flowlines){
+  AOI.type = unique(sf::st_geometry_type(flowlines))
+  if(!AOI.type %in% c("MULTILINESTRING", "LINESTRING")){
+    stop("outlets can only be extracted from flowlines", call. = FALSE)
+  } else {
+    flowlines$geometry = suppressWarnings(
+      flowlines$geometry %>%
+        sf::st_cast("LINESTRING") %>%
+        sf::st_transform(5070) %>%
+        sf::st_line_sample(sample = 1) %>%
+        sf::st_cast("POINT") %>%
+        sf::st_transform(st_crs(flowlines)) %>%
+        sf::st_zm()
+    )
+
+    return(flowlines)
+
+  }
+}
+
+#' @title Find National Hydrography Subsets (Multirealization)
+#' @description Subsets NHD features by location (POINT), area (POLYGON), or set of COMIDs. Mutli realizations are supported allowing you to query for flowlines, catchments, or outlets.
+#' @param AOI an 'area of interest' object (sf POINT or POLYGON)
+#' @param comid Search for NHD features by COMID
+#' @param nwis Search for NHD features by colocated NWIS identifiers
+#' @param streamorder filter returned NHD by streamorder. Returns are for given value and higher. Only usable with AOI, not with comid or nwis inputs
+#' @param realization what realization to return. Default is flowline, options include: outlet, flowline, catchment, and all
+#' @return a single, or list, or simple feature objects
+#' @examples
+#'  point <- sf::st_sfc(sf::st_point(c(-119.845, 34.4146)), crs = 4326)
+#'  find_nhd(point)
+#'  find_nhd(point, realization = "catchment")
+#'  find_nhd(point, realization = "all")
+
+#'  find_nhd(comid = 101)
+#'  find_nhd(nwis  = c(11120000, 11120500))
+
+#'  area <- sf::st_as_sfc(sf::st_bbox(
+#'    c(xmin = -119.8851, xmax =-119.8361, ymax = 34.42439, ymin = 34.40473), crs = 4326))
+
+#'  find_nhd(area)
+#'  find_nhd(area, realization = "flowline", streamorder = 3)
+#' @importFrom methods is
+#' @export
+
+find_nhd <- function(AOI = NULL, comid = NULL, nwis = NULL,
+                     realization = "flowline",
+                     streamorder = NULL){
+
+  if(!is.null(AOI)){
+    if(!methods::is(AOI,c("sf", 'sfc'))){
+      stop("AOI must be of class sf.",.call = FALSE)
+    }
+
+    if(st_geometry_type(AOI) == "POINT"){
+      coords = sf::st_coordinates(AOI)
+      comid  = dataRetrieval::findNLDI(location = c(X = coords[1], Y = coords[2]))$origin$identifier
+      AOI = NULL
+    }
+  }
+
+  if(!is.null(AOI) & !is.null(c(nwis, comid))){
+    stop("Either IDs (comid, nwis) or a spatial AOI can be passed.",.call = FALSE)
+  }
+
+  if(is.null(AOI) & is.null(c(nwis, comid))){
+    stop("IDs (comid, nwis) or a spatial AOI must be passed.",.call = FALSE)
+  }
+
+  good.realizations = c("flowline", "catchment", 'outlet')
+  if(realization == "all"){ realization = good.realizations}
+  if(any(!realization %in% good.realizations)){
+    stop(paste(realization, "not valid.\n Select from", paste(good.realizations, collapse = ", ")))
+  }
+
+
+  catch <- fl <- outlet <-  NULL
```

Comment:
I've been trying to use `.data$attribtue` rather than this method to avoid warnings.

Reply:
changed approach to just populating a list